drivers/coredump/coredump_impl.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Meta Platforms, Inc. and its affiliates.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <debug/coredump.h>
+#include <drivers/coredump.h>
+
+#define DT_DRV_COMPAT zephyr_coredump
+
+struct coredump_config {
+	/* Length of memory_regions array */
+	int length;
+
+	/* Memory regions specified in device tree */
+	int memory_regions[];
+};
+
+struct coredump_data {
+	/* Memory regions registered at run time */
+	sys_slist_t region_list;
+};
+
+static void coredump_impl_dump(const struct device *dev)
+{
+	const struct coredump_config *config = dev->config;
+
+	/*
+	 * Add each memory region specified in device tree to the core dump,
+	 * the memory_regions array should contain two entries per region
+	 * containing the start address and size.
+	 */
+	if ((config->length > 0) && ((config->length % 2) == 0)) {
+		for (int i = 0; i < config->length; i += 2) {
+			uintptr_t start_address = config->memory_regions[i];
+			size_t size = config->memory_regions[i+1];
+
+			coredump_memory_dump(start_address, start_address + size);
```

Comment:
this is not going to work for MMU-enabled platforms because you are getting the phys from the DT and you are trying to dump that directly. You should at least trying to MMU-map the address before calling the dump function.

Reply:
I would suggest 2 possible cases for the dt bindings (as enums)
1. simple static address + size (where the coredump subsys just does a memcpy)
2. size + other device with phandle registers a `dump()` callback (would require early init for coredump devices)